### ISSUE TITLE
Ladder input feature?

### ISSUE THREAD

**Issue by bsteuber** [gcp/leela-zero#1754]
We know that the original AlphaGo had a ladder feature in the network inputs, which worked out really well. For the Zero version, they got rid of it - not because it works better, but because it reads more impressive to use absolutely no human knowledge.

After talking to some Deepmind people, I am quite certain that reintroducing this feature to LZ would get rid of all ladder problems once at for all, so as someone using LZ a lot for analysis (in fact, I just opened a commercial LZ cloud analysis service) I would really love a version with this feature.

I understand that this isn't part of the original project goal (after all, it would have to be renamed “Leela One“). But it seems several people here are willing to experiment with forks and have access to a lot of computing power. So I hope this post might inspire one of you to try this out.

**Comment by gcp** [comment]
Out of scope for this project.

**Comment by gcp** [comment]
There's several people who already made a merger of LZ with Ray's ladder detection AFAIK.

That said I haven't seen LZ fall for a ladder in ages, unless we're talking about a lost position, in which case a ladder is really the least of the problems you'll see.

**Comment by bsteuber** [comment]
Just as an example of ladders still being a big problem for LZ, take https://zero.sjeng.org/viewmatch/28ccb9f11e068608c208f015b1ad18f720ecbdd972101dbfc0828b3f49b86c7f?viewer=wgo from reddit. Its upper right corner variation was played in a lot of other recent games, too (without the ladder working).
When analyzing the position with Lizzie on more or less current next branch and current 20b network, I get these screenshots:

**After >20k playouts:**
![ladder_1](https://user-images.githubusercontent.com/213112/44426195-48caf600-a58e-11e8-8d37-edf04468e546.png)

**After >50k playouts:**
![ladder_2](https://user-images.githubusercontent.com/213112/44426202-4d8faa00-a58e-11e8-900a-174c59bd8042.png)

So, escaping the ladder is a winning move, but even with it being just one move ahead, LZ needs >50k playouts to find it. Black played a losing variation but got unpunished. 

I believe things like that happen far more more open than noticed....

**Comment by gcp** [comment]
Oh, so it's not playing bad ladder at all? It just needs search to overcome a low prior for escaping that way.

**Comment by gcp** [comment]
This kind of thing won't be fixed by ladder detection at all.

**Comment by bsteuber** [comment]
The network has a bias against trying to escape from ladders. That's because in Go the penalty for playing escape moves in non-working ladders is much bigger than the penalty of playing attack-moves in a non-working ladder. Similarly, it has a bias towards attacking stones in ladders. 

A network that just "knows" if a ladder works or not would quickly learn to be heavily biased towards working land against non-working ladders for both sides. Therefore, it would have found the white escaping move already after one visit, then sticking to it. Also, it would never have chosen this variation for black in the first place no matter what visit count.

**Comment by Ttl** [comment]
I have LZ branch with NN feature planes include ladder escape and ladder capture planes and some NN architecture modifications (https://github.com/Ttl/leela-zero/tree/liberty_planes_se2).

On this move ladder escape plane is set on this location:

```
showladders

   a b c d e f g h j k l m n o p q r s t 
19 . . . . . . . . . . . . . . . . . . . 19
18 . . . . . . . . . . . . . . . . . . . 18
17 . . . . . . . . . . . . . . . . . . . 17
16 . . . + . . . . . + . . . . . + . . . 16
15 . . . . . . . . . . . . E . . . . . . 15
14 . . . . . . . . . . . . . . . . . . . 14
13 . . . . . . . . . . . . . . . . . . . 13
12 . . . . . . . . . . . . . . . . . . . 12
11 . . . . . . . . . . . . . . . . . . . 11
10 . . . + . . . . . + . . . . . + . . . 10
 9 . . . . . . . . . . . . . . . . . . .  9
 8 . . . . . . . . . . . . . . . . . . .  8
 7 . . . . . . . . . . . . . . . . . . .  7
 6 . . . . . . . . . . . . . . . . . . .  6
 5 . . . . . . . . . . . . . . . . . . .  5
 4 . . . + . . . . . + . . . . . + . . .  4
 3 . . . . . . . . . . . . . . . . . . .  3
 2 . . . . . . . . . . . . . . . . . . .  2
 1 . . . . . . . . . . . . . . . . . . .  1
   a b c d e f g h j k l m n o p q r s t 
```

It does find the ladder move in 800 visits even though it has very low initial policy score:

```
NN eval=0.553738
Playouts: 511, Win: 61.25%, PV: N15 M17 N14 Q14 Q13 R15 P14 Q15 R13 R17 R18 S18 S17 S16 R16 O14 O13 R17 P16

 N15 ->     619 (V: 68.84%) (N:  2.26%) PV: N15 M17 N14 Q14 Q13 R15 P14 Q15 R13 R17 R18 S18 S17 S16 R16 O14 O13 R17 T17 N13
 R16 ->     148 (V: 50.34%) (N: 73.93%) PV: R16 N15 N3 G3 B3 B2 D9 C11 L17
 P16 ->      24 (V: 49.58%) (N: 14.21%) PV: P16 P15 Q15 N15 Q13 N4 N5 M5
  F3 ->       8 (V: 50.99%) (N:  3.64%) PV: F3 G2 R16 N15 C17
8.6 average depth, 21 max depth
577 non leaf nodes, 1.38 average children
800 visits, 263557 nodes, 799 playouts, 215 n/s
```

I tested before that 64x10 network with same resnet architecture but with additional feature planes has 80% winrate against 64x10 network trained on same data for same amount of steps using same learning rate schedule, so they are useful at least with small models:

```
lz_liberty_planes v lz_next (400/400 games)
board size: 19   komi: 7.5
                    wins              black          white        avg cpu
lz_liberty_planes    329 82.25%       164 82.00%     165 82.50%    134.58
lz_next               71 17.75%       35  17.50%     36  18.00%    125.16
                                      199 49.75%     201 50.25%
```

I didn't have that good success with larger models.

**Comment by bsteuber** [comment]
I guess it takes a lot of self-play training to re-learn all the ladder situations after the feature was added.

**Comment by d7urban** [comment]
Might be an artifact from ELF. I’ve seen several games with this variation
where ELF (both version), never finds that escaping the ladder works, and I
let it search for 1M playouts.

**Comment by alreadydone** [comment]
@Ttl So the training code in your branch can automatically compute and add ladder and liberty features to the current-format training data, right? If so can you point out where the code is? Great work!
@NhanHo Want to see if you can train an even stronger 64x5 with ladder and liberty inputs? If so it would be very useful on mobile phones.

**Comment by Ttl** [comment]
No, the training code doesn't add feature planes. They need to be added in similar way as how supervised training works. Code is in `Training.cpp`.

**Comment by alreadydone** [comment]
So the add_features GTP command of your leelaz takes a sgf (possibly with many games) and output training data with ladder/liberty input planes but without search probabilities (only the move actually played). It would be better to take training data as input ... does leelaz contain a function to convert back the hexadecimal strings encoding the board positions? The training code should have it.
But @NhanHo actually trained his super-strong 64x5 with supervised learning, so the current add_features is probably good enough. But I guess the search probabilities don't hurt; after 3200 visits the visit counts of non-working ladder moves should not be high though the policy priors may be high.

**Comment by Ttl** [comment]
add_features takes training data as input and outputs training data with feature planes.

**Comment by alreadydone** [comment]
That's great; I saw `add_features(sgfname, outname)` in GTP.cpp but missed that the arguments of add_features in its definition actually have `training_file`.

**Comment by AncalagonX** [comment]
@Ttl I've been training my own networks for fun for a while, and finally got around to trying your `liberty_planes_se2` branch today.

It easily built into a working Windows binary, but there are issues with `tfprocess.py`. It looks like you possibly uploaded a mid-merge version to your repository. Check out line # 178 on https://github.com/Ttl/leela-zero/blob/liberty_planes_se2/training/tf/tfprocess.py#L178

Can you also upload a working example v502 weights.txt file, so I can run leelaz.exe to use `add_features`? Thanks.
